motoman_driver/include/motoman_driver/simple_message/motoman_select_tool.h

```diff
@@ -0,0 +1,212 @@
+ *
+ * The byte representation of a select tool command is as follows
+ * (in order lowest index to highest). The standard sizes are given,
+ * but can change based on type sizes:
+ *
+ *   member:             type                                      size
+ *   group_number        (industrial::shared_types::shared_int)    4  bytes
+ *   tool_number         (industrial::shared_types::shared_int)    4  bytes
+ *   sequence_number     (industrial::shared_types::shared_int)    4  bytes
+ *
+ * Note: sequence_number is marked as optional by MotoROS. It will be included
+ * in the SimpleMessage service request, but does not have to be set to any
+ * specific value. It can be used as a "tracking number" that will be "echoed
+ * back in the response message" (according to MotoROS source code comments).
+ *
+ * THIS CLASS IS NOT THREAD-SAFE
+ *
+ */
+
+class SelectTool : public industrial::simple_serialize::SimpleSerialize
+{
+public:
+  /**
+   * \brief Default constructor
+   *
+   * This method creates empty data.
+   *
+   */
+  SelectTool(void);
+
+  /**
+   * \brief Destructor
+   *
+   */
+  ~SelectTool(void);
+
+  /**
+   * \brief Initializes a empty select tool command
+   *
+   */
+  void init();
+
+  /**
+   * \brief Initializes a complete select tool command
+   *
+   * Note: sequence_number default value of 0.
+   */
+  void init(industrial::shared_types::shared_int group_number,
+    industrial::shared_types::shared_int tool_number,
+    industrial::shared_types::shared_int sequence_number = 0);
+
+  /**
+   * \brief Sets group number
+   *
+   * \param group_number Controller group the tool file is defined for
+   */
+  void setGroupNumber(industrial::shared_types::shared_int group_number)
+  {
+    this->group_number_ = group_number;
+  }
+
+  /**
+   * \brief Returns the group number
+   *
+   * \return group number
+   */
+  industrial::shared_types::shared_int getGroupNumber()
+  {
+    return this->group_number_;
+  }
+
+  /**
+   * \brief Sets tool number
+   *
+   * \param tool_number Tool file identifier
+   */
+  void setToolNumber(industrial::shared_types::shared_int tool_number)
+  {
+    this->tool_number_ = tool_number;
+  }
+
+  /**
+   * \brief Returns the tool file identifier
+   *
+   * \return tool number
+   */
+  industrial::shared_types::shared_int getToolNumber()
+  {
+    return this->tool_number_;
+  }
+
+  /**
+   * \brief Sets sequence number
+   *
+   * \param sequence_number Optional sequence number
+   */
+  void setSequenceNumber(industrial::shared_types::shared_int sequence_number)
+  {
+    this->sequence_number_ = sequence_number;
+  }
+
+  /**
+   * \brief Returns the optional sequence number
+   *
+   * \return sequence number
+   */
+  industrial::shared_types::shared_int getSequenceNumber()
+  {
+    return this->sequence_number_;
+  }
+
+  /**
+   * \brief Copies the passed in value
+   *
+   * \param src (value to copy)
+   */
+  void copyFrom(SelectTool &src);
```

Comment:
```suggestion
  void copyFrom(const SelectTool &src);
```

Why not just make a copy constructor instead?

Reply:
Because cargo-cult, and copy-paste, search-and-replace ;)

Perhaps @JeremyZoss can remember.

Reply:
I think most of this is just following the code pattern Shaun used in his original implementation (see [joint_data.h](https://github.com/ros-industrial/industrial_core/commit/3e065210cd8c428dffd94f23dff18fd4b6f79f75#diff-940cfbba5085b9567c89a09d37d661d8c7f5948e29e815904326cff4c8321fb1R129) from 2012), for code consistency.  Some of that design pattern was based on using literally the same code base for ROS and MotoROS `simple_message` classes, which inherits some limitations from MotoROS's C++ implementation.  But that code-sharing is not used in current drivers.  simple_message could probably use some cleanup, but I think it would make sense to tackle that globally (for consistency).  In my opinion, best to keep the style of this PR to match the style of the rest of simple_message.

Reply:
It's a wonder he was even able to use C++. I don't believe C++ is supported officially by MotoPlus.

Reply:
@JeremyZoss wrote:
>  In my opinion, best to keep the style of this PR to match the style of the rest of simple_message.

would that be acceptable to you @marip8 ?

Reply:
That's fine with me. Passing arguments by non-const reference unnecessarily is generally not a great idea, so I just wanted to make a note

Reply:
> Passing arguments by non-const reference unnecessarily is generally not a great idea, so I just wanted to make a note

Yes, agreed.

And please keep doing that. In-depth reviews are very much appreciated.